src/frontend/Ast_to_Mir.ml

```diff
@@ -97,60 +97,120 @@ let neg_inf =
+      || List.exists
+           ~f:(fun (i : Ast.typed_expression) ->
+             UnsizedType.is_container i.emeta.type_ )
+           ast_args
+    then e
+    else
+      (* Container y but scalar args - need to multiply by size(y) *)
+      let trans_ast_obs = trans_expr ast_obs in
+      let type_ = {trans_ast_obs.meta with type_= UnsizedType.UReal} in
+      Expr.Helpers.binop e Times
+        (Expr.Helpers.internal_funapp FnLength [trans_ast_obs] type_) in
   match t with
   | Ast.NoTruncate -> []
   | TruncateUpFrom lb ->
       let fk, fn, tp = find_function_info ccdf_suffices in
-      [ trunc Less lb
-          (targetme lb.emeta.loc
-             (funapp lb.emeta fk fn (inclusive_bound tp lb :: ast_args)) ) ]
+      let lb = trans_expr lb in
+      [ trunc Less "min" lb
+          (targetme lb.meta.loc
+             (size_adjust
+                (funapp lb.meta fk fn
+                   (inclusive_bound tp lb :: trans_exprs ast_args) ) ) ) ]
   | TruncateDownFrom ub ->
       let fk, fn, _ = find_function_info cdf_suffices in
-      [ trunc Greater ub
-          (targetme ub.emeta.loc (funapp ub.emeta fk fn (ub :: ast_args))) ]
+      let ub = trans_expr ub in
+      [ trunc Greater "max" ub
+          (targetme ub.meta.loc
+             (size_adjust (funapp ub.meta fk fn (ub :: trans_exprs ast_args))) )
+      ]
   | TruncateBetween (lb, ub) ->
       let fk, fn, tp = find_function_info cdf_suffices in
-      [ trunc Less lb
-          (trunc Greater ub
-             (targetme ub.emeta.loc
-                (funapp ub.emeta (Ast.StanLib FnPlain) "log_diff_exp"
-                   [ funapp ub.emeta fk fn (ub :: ast_args)
-                   ; funapp ub.emeta fk fn (inclusive_bound tp lb :: ast_args)
-                   ] ) ) ) ]
+      let lb, ub = (trans_expr lb, trans_expr ub) in
+      let expr args =
+        funapp ub.meta (Ast.StanLib FnPlain) "log_diff_exp"
+          [ funapp ub.meta fk fn (ub :: args)
+          ; funapp ub.meta fk fn (inclusive_bound tp lb :: args) ] in
+      let statement =
+        match
+          List.find
+            ~f:(fun (i : Ast.typed_expression) ->
+              UnsizedType.is_container i.emeta.type_ )
+            ast_args
+        with
+        (* If any of the arguments (besides the data) are vectors, need to generate a loop
+           This can go away if https://github.com/stan-dev/stan/issues/1154 is implemented
+        *)
+        | Some e ->
+            let e = trans_expr e in
+            let bound =
+              Expr.Helpers.internal_funapp FnLength [e]
+                {e.meta with type_= UnsizedType.UInt} in
+            let bodyfn (idx : Expr.Typed.t) =
+              let ast_args = trans_exprs ast_args in
+              let args =
+                List.map
+                  ~f:(fun (e : Expr.Typed.t) ->
+                    if UnsizedType.is_container e.meta.type_ then
+                      Expr.Helpers.add_int_index e (Index.Single idx)
```

Comment:
`add_int_index` expects a `Var` or `Indexed` so code like `y ~ normal(a*x + b, sigma)T[0,1];` blows up. You need temporaries to store the arguments because you certainly don't want to recompute the full vector `a*x + b` for each iteration of the loop.

Reply:
Yep, good catch again.

In theory we could store temporaries for this entire operation, since even without the loop it would still be computing that expression ~twice (once for the lupdf, once for the cdf)?

Reply:
Yes, that makes sense.